Allow falling back to single successor meters in grid formulas

shsms opened the pull request:
With this, grid meters that have a single meter connected to them can use that meter as a fallback component.

shsms committed:
Add `meter_fallback_for_meters` parameter to FallbackExpr generator

When true, fallback expressions will consider (single) successor
meters as fallback components for meters.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Allow falling back to single successor meters in grid formulas

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update release notes

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

nhatcher-frequenz commented on src/graph/formulas/fallback.rs:
```diff
@@ -20,9 +20,11 @@ where
         &self,
         component_ids: impl IntoIterator<Item = u64>,
         prefer_meters: bool,
+        meter_fallback_for_meters: bool,
```
Nit: It's a bit hard to understand what these two parameters mean. Ad things like:

```rust
self.graph.fallback_expr(self.pv_inverter_ids, false, false)
```
make it hard to understand what each false refers to.  Since it is just two booleans Maybe we can use an enum with four variants or two enums with two variants each (so that it would be a type error in you mix the variants)?

shsms replied:
True, but they are separate config params, not sure if we should combine them like that.  It also gets hard if we want to add a couple of more bools. (which would happen, because we're planning to introduce formula-level configs soon)

Maybe a struct would be better?

shsms replied:
Although it also feels overkill when there's only two.

shsms replied:
Maybe switch to two enums, each with two variants, replacing the bools.  Ok, let me try that.

shsms replied:
Actually I could just inline the `fallback_expr` method everywhere, then drop it, so that everywhere it would look like:

```
        let expr = FallbackExpr {
            prefer_meters: true,
            meter_fallback_for_meters: true,
            graph: &graph,
        }
        .generate(BTreeSet::from_iter(vec![1]))?;
```

This feels less clunky than having a config struct and is very readable.  Only the tests will get a bit verbose.  Otherwise there are very few uses, so it should be fine.

shsms replied:
Refactoring further, such that graph goes in the generate call.

shsms replied:
Ok, ready again.

shsms replied:
Actually no, can cleanup further. :D doing now.

shsms replied:
Ok, really ready now.

shsms committed:
Refactor to not store a graph reference in `FallbackExpr`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Inline all uses of `fallback_expr`

The `fallback_expr` function was supposed to be a easy to use
constructor for the FallbackExpr and a call to `generate` on it.

But creating the struct directly specifying the field names makes it
easier to read.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove unused `fallback_expr` function

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

matthias-wende-frequenz commented on src/graph/formulas/fallback.rs:
```diff
@@ -67,9 +70,14 @@ where
+        let has_successor_meters = self.graph.has_meter_successors(component_id)?;
+
+        if !self.meter_fallback_for_meters && has_successor_meters {
+            return Ok(Some(Expr::component(component_id)));
```
Interesting. In the old version when the successor was a meter the return value was `Ok(None)` and now it changed to `Ok(Some(Expr::component(component_id)))` (when we aren't looking for fallback meters). Can you please explain why this changed?

matthias-wende-frequenz replied:
It looks like you have been fixing a logical bug here. Is my understanding correct?

shsms replied:
It wasn't a bug, but the behaviour was different earlier:

```
        if !component.is_meter() || self.graph.has_meter_successors(component_id)? {
            return Ok(None);
        }
```

This meant that a meter fallback formula was not possible at that level, so it tried as a component, and if that fails too, then use the component in the `else` block here: https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/9d70e58f25e7920c9b31bb11c8d8dd49ddb6d70f/src/graph/formulas/fallback.rs#L55-L61

Now this has changed to:

```
        if !component.is_meter() {
            return Ok(None);
        }
```

And for the case with subsequent meters, we need more handling, and for the case where we don't want further search, we return the component from here, so it wouldn't enter the `else` block above.

So yes, I guess it improves logical correctness and modularity, but no change in behaviour at the end.
